test_backend.py

```diff
@@ -117,13 +123,201 @@ def test_robot_is_repaired(damages_before, tile, damages_after):
+                        ])
+def test_robot_changed_direction(direction_before, tile, direction_after):
+    """
+    When robot is on GearTile, he should be rotated according to the direction of the tile.
+    Check that his direction changed after applying tile effect.
+    """
+    robot = Robot(direction_before, None, None, (0, 0))
+    state = State({(0, 0): [tile]}, [robot], 1)
+    apply_tile_effects(state)
+    assert robot.direction == direction_after
+
+
+# HoleTile hidden in walk method
+
+@pytest.mark.parametrize(("lives_before", "lives_after"),
+                        [(3, 2),
+                         (2, 1),
+                         (1, 0),
+                        ])
+def test_robot_died(lives_before, lives_after):
+    """
+    When robot comes to a HoleTile (or goes / is pushed out of the game board),
+    he gets killed.
+    Check that his lives were lowered, he got inactive till the next game round
+    and his coordinates changed to the (-1, -1).
+    """
+    robot = Robot(Direction.N, None, None, (0, 0))
+    state = State({(0, 1): [HoleTile(None, None, None)]}, [robot], 1)
+    robot.lives = lives_before
+    robot.walk(1, state)
+    assert robot.lives == lives_after
+    assert robot.inactive == True
+    assert robot.coordinates == (-1, -1)
+
+
+# FlagTile
+
+@pytest.mark.parametrize(("flags_before", "tile", "flags_after", "start_coordinates_after"),
+                        [(3, FlagTile(None, None, [{'value': 1}]),  3,  (1, 1)),
+                         (3, FlagTile(None, None, [{'value': 4}]),  4, (0, 0)),
+                         (3, FlagTile(None, None, [{'value': 5}]),  3, (1, 1)),
+                        ])
+def test_robot_collected_flags(flags_before, tile, flags_after, start_coordinates_after):
+    """
+    When a robot stands on FlagTile with appropriate number (+1 to his current flag count), he collects it.
+    He doesn't collect the flags with the other number than defined.    They don't have any effect on him.
+    With the flag collected the starting coordinates change to the tile's coordinates.
+    """
+    robot = Robot(None, None, None, (0, 0))
+    state = State({(0, 0): [tile]}, [robot], 1)
+    robot.flags = flags_before
+    robot.start_coordinates = (1, 1)
+    apply_tile_effects(state)
+    assert robot.flags == flags_after
+    assert robot.start_coordinates == start_coordinates_after
+
+
+# WallTile
+
+@pytest.mark.parametrize(("input_coordinates", "output_coordinates"),
+                        [((0, 0), (0, 0)),
+                         ((1, 0), (1, 0)),
+                         ((2, 0), (2, 1)),
+                         ((3, 0), (3, 1)),
+                         ])
+def test_robot_is_stopped_by_wall(input_coordinates, output_coordinates):
+    """
+    Take robot's coordinates, move's direction and distance and assert robot
+    was moved to correct coordinates.
+    A special map test_walls was created in order to test this feature.
```

Comment:
Přiznám se, že mě v té mapě zmátly startovací políčka, která pak nepoužíváš... Pokud jsou tam k tomu, aby si tu mapu mohl člověk pustit přes module `game.py`, tak ok.

Reply:
Ano a ne, příznávám, že je to trochu nekonzistentní (laser mapa nemá startovací políčka). Protože v parametrů stejně určuji robotům startovací políčka, nejsou třeba v mapě. Původně jsem chtěla využít pouze mapu, ale protože nemůžu jistě určit pořadí robotů v seznamu robotů, test mi randomově failoval. Radši tu mapu upravím, aby nemátla.